tedana/decomposition/gift.py

```diff
@@ -0,0 +1,612 @@
+
+    # Index vectors
+    k = np.arange(-(n_points - 1) / 2, ((n_points - 1) / 2) + 1)
+    k1 = k[k < -(n_points - 1) / 4]
+    k2 = k[abs(k) <= (n_points - 1) / 4]
+
+    # Equation 37 of [1]: window defined in three sections
+    parzen_w1 = 2 * (1 - abs(k1) / (n_points / 2))**3
+    parzen_w2 = 1 - 6 * (abs(k2) / (n_points / 2))**2 + 6 * (abs(k2) / (n_points / 2))**3
+    parzen_w = np.hstack((parzen_w1, parzen_w2, parzen_w1[::-1])).T
+
+    return parzen_w
+
+
+def _ent_rate_sp(data, sm_window):
+    """
+    Calculate the entropy rate of a stationary Gaussian random process using
+    spectrum estimation with smoothing window.
+
+    Parameters
+    ----------
+    data : ndarray
+        Data to calculate the entropy rate of and smooth
+    sm_window : boolean
+        Whether there is a Parzen window to use
+
+    Returns
+    -------
+    ent_rate : float
+        The entropy rate
+
+    Notes
+    -----
+    This function attempts to calculate the entropy rate following (Li et al., 2007)
+    """
+
+    dims = data.shape
+
+    # Normalize x_sb to be unit variance
+    data_std = np.std(np.reshape(data, (-1, 1)))
+
+    # Make sure we do not divide by zero
+    if data_std == 0:
+        raise ValueError('Divide by zero encountered.')
+    data = data / data_std
+
+    if sm_window:
+
+        M = [int(i) for i in np.ceil(np.array(dims) / 10)]
+
+        # Get Parzen window for each spatial direction
+        if data.ndim >= 3:
+            parzen_w_3 = np.zeros((2 * dims[2] - 1, ))
+            parzen_w_3[(dims[2] - M[2] - 1):(dims[2] + M[2])] = _parzen_win(2 * M[2] + 1)
+
+        if data.ndim >= 2:
+            parzen_w_2 = np.zeros((2 * dims[1] - 1, ))
+            parzen_w_2[(dims[1] - M[1] - 1):(dims[1] + M[1])] = _parzen_win(2 * M[1] + 1)
+
+        if data.ndim >= 1:
+            parzen_w_1 = np.zeros((2 * dims[0] - 1, ))
+            parzen_w_1[(dims[0] - M[0] - 1):(dims[0] + M[0])] = _parzen_win(2 * M[0] + 1)
+
+    if data.ndim == 2 and min(dims) == 1:
+        # Apply window to 1D
+        data_corr = _autocorr(data)
+        data_corr = data_corr * parzen_w_1
+        data_fft = fftshift(fft(data_corr))
+
+    elif data.ndim == 2 and min(dims) != 1:
+        # Apply windows to 2D
+        data_corr = _autocorr(data)
+
+        # Create bias-correcting vectors
+        v1 = np.hstack((np.arange(1, dims[0] + 1),
+                        np.arange(dims[0] - 1, 0, -1)))[np.newaxis, :]
+        v2 = np.hstack((np.arange(1, dims[1] + 1),
+                        np.arange(dims[1] - 1, 0, -1)))[np.newaxis, :]
+
+        vd = np.dot(v1.T, v2)
+
+        # Bias-correct
+        data_corr /= vd
+
+        # Apply 2D Parzen Window
+        parzen_window_2D = np.dot(parzen_w_1, parzen_w_2.T)
+        data_corr = data_corr * parzen_window_2D
+        data_fft = fftshift(fft2(data_corr))
+
+    elif data.ndim == 3 and min(dims) != 1:
+        # Apply windows to 3D
+        # TODO: replace correlate2d with 3d if possible
```

Comment:
Can be removed, unless extending `fftconvolve` to 3D makes sense.

Reply:
There are some other if statements that could also me removed. I've go through the whole thing again and see where it makes sense to remove them.